src/TimeoutMigrationTool/ASQ/ASQEndpointMigrator.cs

```diff
@@ -0,0 +1,111 @@
+﻿namespace Particular.TimeoutMigrationTool.ASQ
+{
+    using System;
+    using Microsoft.Azure.Cosmos.Table;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Cosmos.Table.Queryable;
+
+    public class ASQEndpointMigrator : ITimeoutsTarget.IEndpointTargetBatchMigrator
+    {
+        public ASQEndpointMigrator(CloudTableClient client, string destinationTimeoutTableName, string stagingTableName)
+        {
+            this.client = client;
+            this.destinationTimeoutTableName = destinationTimeoutTableName;
+            this.stagingTableName = stagingTableName;
+        }
+
+        public async ValueTask<int> StageBatch(IReadOnlyList<TimeoutData> timeouts, int batchNumber)
+        {
+            await PurgeStagingTable();
+
+            var stagedTimeouts = timeouts.Select(timeout => StagedDelayedMessageEntity.FromTimeoutData(timeout, batchNumber)).GroupBy(x => x.PartitionKey);
+            var table = client.GetTableReference(stagingTableName);
+
+            var timeoutsStaged = await BatchProcessRecords(table, stagedTimeouts, (operation, entity) => operation.InsertOrReplace(entity));
+            return timeoutsStaged;
+        }
+
+        public async ValueTask<int> CompleteBatch(int batchNumber)
+        {
+            var stagingTable = client.GetTableReference(stagingTableName);
+            var query = stagingTable.CreateQuery<StagedDelayedMessageEntity>().Where(x => x.PartitionKey == batchNumber.ToString()).AsTableQuery();
+            var delayedMessages = await stagingTable.ExecuteQueryAsync(query, int.MaxValue, CancellationToken.None);
```

Comment:
The consequence of using the copied `ExecuteQueryAsync` from ASQ is that we first materialize ALL the data from the server into memory before we execute any other operation. This seems hugely exhaustive and potentially dangerous if many timeouts need to be fetched compared to doing operations per page segment as they stream from the server. That's the reason why we deliberately did not use this method and went for slightly "uglier" code style with the do/while directly in your face where it is needed